psl-core/src/main/java/org/linqs/psl/evaluation/statistics/RankingEvaluator.java

```diff
@@ -21,255 +21,254 @@
-    }
+    /*
+    Grab the Constant argument of an atom in a specified position.
+    */
 
-    public RankingEvaluator(double threshold, RepresentativeMetric representative) {
-        if (threshold < 0.0 || threshold > 1.0) {
-            throw new IllegalArgumentException("Threhsold must be in (0, 1). Found: " + threshold);
-        }
+    public Constant getArgAtPosition(GroundAtom atom, int position) {
+        Constant[] args = atom.getArguments();
+        return args[position];
+    }
 
-        this.threshold = threshold;
-        this.representative = representative;
+    /*
+    Convert the Iterables returned by TrainingMap methods
+    into Lists.
+    */
 
-        truth = new ArrayList<GroundAtom>();
-        predicted = new ArrayList<GroundAtom>();
+    public static <E> ArrayList<E> iterableToList(Iterable<E> iterable) {
+        ArrayList<E> list = new ArrayList<E>();
+        for (E element : iterable) {
+            list.add(element);
+        }
+        return list;
     }
 
     @Override
     public void compute(TrainingMap trainingMap) {
-        compute(trainingMap, null);
+        if (defaultPredicate == null) {
+            throw new UnsupportedOperationException("RankingEvaluators must have a default predicate set (through config).");
+        }
+
+        compute(trainingMap, StandardPredicate.get(defaultPredicate));
     }
 
     @Override
     public void compute(TrainingMap trainingMap, StandardPredicate predicate) {
-        truth = new ArrayList<GroundAtom>(trainingMap.getLabelMap().size());
-        predicted = new ArrayList<GroundAtom>(trainingMap.getLabelMap().size());
-
-        for (Map.Entry<GroundAtom, GroundAtom> entry : getMap(trainingMap)) {
-            if (predicate != null && entry.getKey().getPredicate() != predicate) {
-                continue;
-            }
+        /*
+        Populate lists of all predicted/truth atoms
+        which have the specified predicate.
+        */
 
-            truth.add(entry.getValue());
-            predicted.add(entry.getKey());
-        }
+        predictedAtoms = iterableToList(trainingMap.getAllPredictions());
+        truthAtoms = iterableToList(trainingMap.getAllTruths());
 
-        Collections.sort(truth);
-        Collections.sort(predicted);
-    }
+        /*
+        Filter out atoms from these lists which
+        don't have the specified predicate.
+        */
 
-    @Override
-    public double getRepMetric() {
-        switch (representative) {
-            case AUROC:
-                return auroc();
-            case POSITIVE_AUPRC:
-                return positiveAUPRC();
-            case NEGATIVE_AUPRC:
-                return negativeAUPRC();
-            default:
-                throw new IllegalStateException("Unknown representative metric: " + representative);
-        }
-    }
+        predictedAtoms.removeIf(atom -> atom.getPredicate() != predicate);
```

Comment:
We need to support Java 7 for now, so no lambdas.
(Hopefully we can move up in the summer.)